text/3700-ptr-assume-moved.md

```diff
@@ -0,0 +1,252 @@
+mechanism is intended to enable the use of architecture features such as AArch64 Top-Byte Ignore
+(TBI) to facilitate use-cases such as high-bit pointer tagging.  An example application of this
+mechanism would be writing a tagging memory allocator.
+
+# Motivation
+[motivation]: #motivation
+
+The term "pointer tagging" could be used to mean either high-bit tagging or low-bit tagging.
+Architecture extensions such as AArch64 Top-Byte Ignore make the CPU disregard the top bits of a
+pointer when determining the memory address, leaving them free for other uses.
+
+This RFC is specifically concerned with creating those high-bit tagged pointers for systems which
+can make use of such architecture features. High-bit tagged pointers pose a somewhat tricky
+challenge for Rust, as the memory model still considers those high bits to be part of the address.
+Thus, from the memory model's perspective, changing those bits puts the pointer outside of its
+original allocation, despite it not being the case as far as the hardware & OS are concerned.  This
+makes loads and stores using the pointer Undefined Behaviour, despite the fact that if such loads
+and stores were to be directly done in assembly they would be perfectly safe and valid.
+
+Whenever this RFC refers to a "tagged pointer", it should be taken to mean a pointer that had some
+of its top bits set to non-0 values.
+
+Tagged pointers are pointers in which the unused top bits are set to contain some metadata - the
+tag.  No 64-bit architecture today actually uses a 64-bit address space. Most operating systems only
+use the lower 48 bits, leaving higher bits unused. The remaining bits are for the most part used to
+distinguish userspace pointers (0x00) from kernelspace pointers (0xff), at least on Linux.  Certain
+architectures provide extensions, such as TBI on AArch64, that make it easier for programs to make
+use of those unused bits to insert custom metadata into the pointer without having to manually mask
+them out prior to every load and store. This tagging method can be used without said architecture
+extensions - by masking out the bits manually - albeit said extensions make it more efficient.
+
+Currently, Rust does not support directly using TBI and related architecture extensions that
+facilitate the use of tagged pointers. This could potentially cause issues in cases such as working
+with TBI-enabled C/C++ components over FFI, or when writing a tagging memory allocator. While there
+is no explicit support for this in C/C++, due to there not being Strict Provenance restrictions it
+is straightforward to write a 'correct' pointer tagging implementation by simply doing a `inttoptr`
+cast inside the memory allocator implementation, be it a custom C `malloc` or using a custom C++
+`Allocator`. The goal of this effort is to create a Rust API for implementing this type of
+functionality that is guaranteed to be free of Undefined Behaviour.
```

Comment:
FWIW doing this in the global_allocator in Rust is "probably also fine", same as in C/C++. Those are not inlined and so the compiler won't see the problematic pointer manipulations. The allocator is already special and it could be possible to fold this magic into that existing magic. (It'd still be nicer to do something proper. I just don't want there to be the impression that Rust somehow supports this less than C/C++ do.)

The main case where new language support is required is when doing this inside regular code.